### PR TITLE
Define emscripten exception struct in native code. NFC

### PR DESCRIPTION
**Pull request by sbc100**
This is really step 1 of moving more of this code to be
native.

### CODE REVIEW

File: system/lib/libc/musl/arch/emscripten/atomic_arch.h [hunk 1]
```diff
@@ -36,9 +36,9 @@ static inline void a_or_l(volatile void *p, long v)
 #define a_cas_p a_cas_p
 static inline void *a_cas_p(volatile void *p, void *t, void *s)
 {
-	void* expected = t;
-	__c11_atomic_compare_exchange_strong((_Atomic uintptr_t*)p, &expected, s, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
-	return expected;
+	uintptr_t expected = (uintptr_t)t;
+	__c11_atomic_compare_exchange_strong((_Atomic uintptr_t*)p, &expected, (uintptr_t)s, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+	return (void*)expected;
```
Comment (kripken) [inline system/lib/libc/musl/arch/emscripten/atomic_arch.h:41]: Are these atomic changes related?

File: tools/gen_struct_info.py [hunk 2]
```diff
@@ -141,6 +141,8 @@ def parse_c_output(lines):
 
   for line in lines:
     arg = line[1:].strip()
+    if '::' in arg:
+      arg = arg.split('::', 1)[1]
```
Comment (kripken) [inline tools/gen_struct_info.py:145]: I'm surprised this works... is the namespace used in the file, I guess, so it's optional?

File: tools/system_libs.py [hunk 3]
```diff
@@ -932,6 +932,9 @@ def get_cflags(self):
       cflags.append('-D_LIBCXXABI_NO_EXCEPTIONS')
     elif self.eh_mode == Exceptions.EMSCRIPTEN:
       cflags.append('-D__USING_EMSCRIPTEN_EXCEPTIONS__')
+      # The code used to interpret exceptions during terminate
+      # is not compatible with emscripten exceptions.
+      cflags.append('-DLIBCXXABI_SILENT_TERMINATE')
```
Comment (kripken) [inline tools/system_libs.py:937]: some of these changes appear in another PR, which I guess should land first?

**Review by kripken — APPROVED** [review]

File: system/lib/libc/musl/arch/emscripten/atomic_arch.h (hunk 1 above)
Comment (sbc100) [reply system/lib/libc/musl/arch/emscripten/atomic_arch.h:41]: This is needed so that the header is compatible with C++..  this is musl internal header and it normally nevers goes through the C++ compiler.

I could split `struct_info_internal.json` into two.. one for C headers and one for C++ headers.. but it seemed better/simpler to fix this header to avoid the C++ error.

File: tools/gen_struct_info.py (hunk 2 above)
Comment (sbc100) [reply tools/gen_struct_info.py:145]: This just allows us to write `C_STRUCTS.__cxa_exception` in the JS library code instead of `C_STRUCTS.__cxxabiv1::__cxa_exception`..  as long as there are no collisions it works fine.

File: tools/system_libs.py (hunk 3 above)
Comment (sbc100) [reply tools/system_libs.py:937]: I'm not sure what happened here.   This change was based on #14430 which is now landed.  Maybe i didn't set the correct upstream branch in the UI.
